tests/framework/report.py

```diff
@@ -0,0 +1,264 @@
+    def test_net_device():
+        """
+        Test description
+        @type: regression
+        @issue: https://github.com/firecracker-microvm/firecracker/issues/1
+        """
+
+    return "Actual value: 8", "Maximum value: 10"
+
+On the first line we define the test description. Each subsequent item is
+defined by specifying a token enclosed by '@' and ':'. None of these tokens
+is mandatory to be defined in the docstring. If the token is missing, the
+report may contain the default value or no value at all.
+
+Since we also want to store what test criteria we're targetting for a test to
+pass, the framework picks up the values returned by each test to report this
+data.
+
+The format used for returning the test criteria is a tuple, with the first
+element being the actual value or result and the second one being the checked
+limits.
+
+For example, for the test above the reported criteria would be
+    'Maximum value: 10'
+While the test result would be:
+    'Actual value: 8'
+
+The test framework does not do any comparison here, so it's the users
+responsibility to properly report these values.
+'''
+import csv
+import inspect
+import json
+import re
+import subprocess
+from collections import namedtuple
+from pathlib import Path
+from . import mpsing   # pylint: disable=relative-beyond-top-level
+
+# Try to see if we're in a git repo. If yes, set the commit ID
+COMMIT_ID = ""
+try:
+    COMMIT_ID = subprocess.check_output(
+        "git rev-parse HEAD",
+        shell=True
+    ).decode().strip()
+except subprocess.CalledProcessError:
+    COMMIT_ID = "Out of tree build"
+
+
+class Report(mpsing.MultiprocessSingleton):
+    """
+    Class that holds data about what tests have been executed.
+
+    Generates a test report at the end of the test session.
+    """
+
+    FNAME_JSON = "test_report.json"
+    FNAME_CSV = "test_report.csv"
+
+    # Define a container for representing a test item.
+    # Contains a test item name, the value and optinally defines what values
+    # it accepts.
+    DocItem = namedtuple("DocItem", ["value", "one_of"])
+
+    # This is where we define what items we look-up in the docstring.
+    doc_items = {
+        # What the test does
+        "description": DocItem("", None),
+
+        # If the test passed, failed or was skipped - automatically filled
+        "outcome": DocItem("", None),
+
+        # How long the test took - automatically filled
+        "duration": DocItem(0, None),
+
+        # What kind of test we're running. We only accept a predefined list
+        # of tests.
+        "type": DocItem("", ["build", "functional", "performance",
+                             "security", "style"]),
+
+        # What we take into account to pass a test
+        "criteria": DocItem("", None),
+
+        # Actual result compared to the criteria
+        "result": DocItem("", None),
+
+        # GitHub issue to a link is related to this test
+        "issue": DocItem("", None),
```

Comment:
I would move this under "description" or "type". I think "result" should be the last field.